scrypto/src/core/data.rs

```diff
@@ -0,0 +1,46 @@
+use crate::buffer::*;
+use crate::core::DataAddress;
+use crate::engine::api::RadixEngineInput::WriteData;
+use crate::engine::call_engine;
+use sbor::rust::cell::RefMut;
+use sbor::rust::ops::{Deref, DerefMut};
+use sbor::Encode;
+
+pub struct DataRef<'a, V: Encode> {
+    pub value: RefMut<'a, V>,
+}
+
+impl<'a, V: Encode> Deref for DataRef<'a, V> {
+    type Target = V;
+
+    fn deref(&self) -> &Self::Target {
+        self.value.deref()
+    }
+}
+
+pub struct DataRefMut<'a, V: Encode> {
+    pub address: DataAddress,
+    pub value: RefMut<'a, V>,
+}
+
+impl<'a, V: Encode> Drop for DataRefMut<'a, V> {
+    fn drop(&mut self) {
+        let bytes = scrypto_encode(self.value.deref());
+        let input = WriteData(self.address.clone(), bytes);
+        let _: () = call_engine(input);
+    }
+}
+
+impl<'a, V: Encode> Deref for DataRefMut<'a, V> {
+    type Target = V;
+
+    fn deref(&self) -> &Self::Target {
+        self.value.deref()
+    }
+}
+
+impl<'a, V: Encode> DerefMut for DataRefMut<'a, V> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.value.deref_mut()
+    }
+}
```

Comment:
Why not `Ref` and `RefMut` directly? 

With this extra layer of abstraction, we lost all the traits implemented by `Ref/RefMut`.  Currently unable to format a data ref.

Reply:
Wouldn't be able to use the `drop` and flush write mechanism with `Ref/RefMut`.

Reply:
Cool. Let's at least implement the Debug and Fmt trait like `Ref/RefMut`.